### ISSUE TITLE
How to obtain binary contact maps?

### ISSUE THREAD

**Issue by enriccorona** [samarth-robo/contactdb_utils#1]
Hi! Congratulations for your paper and nice work, the idea of using thermal cameras to find grasp points is very original!

I am trying to work on the 3D prediction of contact maps, and I was wondering how you binarize the labels to train and evaluate your models. From the paper, I understand you use a sigmoid on the temperature values and threshold at 0.4.  I downloaded the contact maps and I am not sure if the sigmoid is applied in it already. But thresholding at 255*0.4=102, gives ~65% of vertices are positive (handoff), which seems way too much.

If those don't contain the sigmoid, I should run the same function as in "scripts/paper_scripts/proc_thermal_images.py"? I tried with this with the mug object and most meshes (handoff) have between 3% and 10% of contact vertices, which seems reasonable to me.

Thanks

**Comment by samarth-robo** [comment]
Hi @enriccorona thanks!
Binarization is only needed for prediction, so you'll find that code in the [companion repository](https://github.com/samarth-robo/contactdb_prediction) which has all the ML code. Specifically, the function `discretize_texture()` in [utils.py](https://github.com/samarth-robo/contactdb_prediction/blob/diversenet/utils.py).

I've included the `scripts/show_contactmap.py` script in this repository for viewing the contact maps. The easiest option would be to copy-paste the `discretize_texture()` function in there, and call it after applying the sigmoid but before applying the inferno color-map, in `texture_proc()`.

The `scripts/paper_scripts/proc_thermal_images.py` file was just used to generate some images for the paper. It uses fixed sigmoid parameters. For the contact maps, you can actually do better -- fit a sigmoid to that specific contact map, apply it (done by `texture_proc()`), and then threshold (done by `discretize_texture()`).

Sorry the code is distributed between two repositories! I thought it would make more sense to have a separate repository for the ML experiments. Let me know if you have more questions!

**Comment by enriccorona** [comment]
Hi @samarth-robo!

Thanks for your answer! In that case, I'm finding that, for instance, mug 3d models have an average of 60% vertices labelled as positive (meaning they are being touched by hand during grasping), for evaluation. I thought it would be less, but I assume that is expected then!

**Comment by samarth-robo** [comment]
@enriccorona Ah I see. `discretize_texture()` sets the value of vertices which were not observed during scanning as `2`, to mark them as "dont-care". You are most likely counting them as contacted.

The loss function ignores those vertices, since we don't know the contact value there since we haven't scanned them at all. Such vertices are few, we tried our best to scan the whole object. But they exist at the bottom and inside of the mug, for example.

I just pushed a commit to `contactdb_utils`, adding a `have_dontcare` switch to `discretize_texture()`. You should call `discretize_texture()` with `have_dontcare=False`. That will give you much more reasonable contact fractions.

Sorry for the late reply!
